### PR TITLE
MONGOCRYPT-549 Unify payload derivations in mongocrypt-marking.c

### PR DESCRIPTION
**Pull request by sgolemon**
Removing more copypasta from a series of methods that do the same thing with minor differences.  In fact, insert_updateForRange is a superset of insert_update.

I wanted to merge the find variants as well, but they diverge too much to make it worth it.

This is in support of having yet more variants come into this file and wanting to keep it as tidy as possible.

### CODE REVIEW

**Review by erwee — APPROVED** [review]

File: src/mongocrypt-marking.c [hunk 1]
```diff
@@ -512,120 +550,134 @@ _mongocrypt_fle2_placeholder_common (_mongocrypt_key_broker_t *kb,
...
+   res = true;
+fail:
+   _mongocrypt_buffer_cleanup (&value);
+   return res;
+}
+
+/**
+ * Payload subtype 4: FLE2InsertUpdatePayload
+ *
+ * {d: EDC, s: ESC, c: ECC, p: encToken, v: value, e: serverToken}
```
Comment (kevinAlbs) [inline src/mongocrypt-marking.c:651]:
```suggestion
 * {d: EDC, s: ESC, c: ECC, p: encToken, u: indexKeyId, t: type, v: value,
 *  e: serverToken}
```

File: src/mongocrypt-marking.c [hunk 2]
```diff
@@ -729,34 +781,30 @@ get_edges (mc_FLE2RangeInsertSpec_t *insertSpec,
    return NULL;
 }
 
-// _mongocrypt_fle2_placeholder_to_insert_update_ciphertextForRange transforms a
-// FLE2EncryptedPlaceholder to an FLE2InsertUpdatePayload for the range
-// algorithm.
+/**
+ * Payload subtype 4: FLE2InsertUpdatePayload for range updates
+ *
+ * {d: EDC, s: ESC, c: ECC, p: encToken, v: value, e: serverToken,
```
Comment (kevinAlbs) [inline src/mongocrypt-marking.c:787]:
```suggestion
 * {d: EDC, s: ESC, c: ECC, p: encToken, u: indexKeyId, t: type, v: value,
 *  e: serverToken,
```

File: src/mongocrypt-marking.c [hunk 3]
```diff
@@ -512,120 +550,134 @@ _mongocrypt_fle2_placeholder_common (_mongocrypt_key_broker_t *kb,
...
 
+// Shared implementation for insert/update and insert/update ForRange
 static bool
-_mongocrypt_fle2_placeholder_to_insert_update_ciphertext (
+_mongocrypt_fle2_placeholder_to_insert_update_common (
    _mongocrypt_key_broker_t *kb,
-   _mongocrypt_marking_t *marking,
-   _mongocrypt_ciphertext_t *ciphertext,
+   mc_FLE2InsertUpdatePayload_t *out,
+   int64_t *contentionFactor,
```
Comment (kevinAlbs) [inline src/mongocrypt-marking.c:558]:
```suggestion
```

Callers do not use the resulting chosen `contentionFactor`.

Suggest also renaming the parameter `maxContentionCounter => contentionFactor` in `_mongocrypt_fle2_placeholder_common`.

File: src/mongocrypt-marking.c [hunk 4]
```diff
@@ -512,120 +550,134 @@ _mongocrypt_fle2_placeholder_common (_mongocrypt_key_broker_t *kb,
...
+          &payload,
+          &contentionFactor,
+          &common,
+          placeholder,
+          &placeholder->v_iter,
+          status)) {
+      goto fail;
    }
 
+   _mongocrypt_ciphertext_init (ciphertext);
```
Comment (kevinAlbs) [inline src/mongocrypt-marking.c:680]:
```suggestion
```

Current callers of `_mongocrypt_marking_to_ciphertext` call `_mongocrypt_ciphertext_init` on the passed `ciphertext`. Suggest removing the `_mongocrypt_ciphertext_init` to not attempt to support uninitialized `ciphertext` from being passed.

Suggest adding a comment to `mongocrypt-marking-private.h` to communicate the expectation:
```c
// Callers are expected to initialize `ciphertext` with _mongocrypt_ciphertext_init before calling.
bool
_mongocrypt_marking_to_ciphertext (void *ctx,
                                   _mongocrypt_marking_t *marking,
                                   _mongocrypt_ciphertext_t *ciphertext,
                                   mongocrypt_status_t *status)
   MONGOCRYPT_WARN_UNUSED_RESULT;
```

File: src/mongocrypt-marking.c [hunk 5]
```diff
@@ -935,7 +893,7 @@ _mongocrypt_fle2_placeholder_to_insert_update_ciphertextForRange (
       }
    }
 
-   // Serialize.
+   _mongocrypt_ciphertext_init (ciphertext);
```
Comment (kevinAlbs) [inline src/mongocrypt-marking.c:896]:
```suggestion
```

Current callers of `_mongocrypt_marking_to_ciphertext` call `_mongocrypt_ciphertext_init` on the passed `ciphertext`. Suggest removing the `_mongocrypt_ciphertext_init` to not attempt to support uninitialized `ciphertext` from being passed.

Suggest adding a comment to `mongocrypt-marking-private.h` to communicate the expectation:
```c
// Callers are expected to initialize `ciphertext` with _mongocrypt_ciphertext_init before calling.
bool
_mongocrypt_marking_to_ciphertext (void *ctx,
                                   _mongocrypt_marking_t *marking,
                                   _mongocrypt_ciphertext_t *ciphertext,
                                   mongocrypt_status_t *status)
   MONGOCRYPT_WARN_UNUSED_RESULT;
```

**Review by kevinAlbs — APPROVED** [review]
The clean-up efforts are much appreciated.

File: src/mongocrypt-marking.c (hunk 1 above)
Comment (sgolemon) [reply src/mongocrypt-marking.c]: Done

File: src/mongocrypt-marking.c [hunk 6]
```diff
@@ -512,120 +550,134 @@ _mongocrypt_fle2_placeholder_common (_mongocrypt_key_broker_t *kb,
...
+ * {d: EDC, s: ESC, c: ECC, p: encToken, v: value, e: serverToken}
+ */
+static bool
+_mongocrypt_fle2_placeholder_to_insert_update_ciphertext (
+   _mongocrypt_key_broker_t *kb,
+   _mongocrypt_marking_t *marking,
+   _mongocrypt_ciphertext_t *ciphertext,
+   mongocrypt_status_t *status)
+{
+   mc_FLE2EncryptionPlaceholder_t *placeholder = &marking->fle2;
```
[the remark this answers is not in the document]
Comment (sgolemon) [reply src/mongocrypt-marking.c:669]: > Put back BSON_ASSERT_PARAM(marking) and BSON_ASSERT_PARAM(ciphertext)

Done, with some other assertions too.

> Should probably also have BSON_ASSERT(placeholder) after this line.

It's a ref to a local, the pointer can't possibly be null for reasons that are local to the assertion.

File: src/mongocrypt-marking.c (hunk 4 above)
Comment (sgolemon) [reply src/mongocrypt-marking.c]: Got it. Yeah, I see the caller now.  Added comment and removed inits here and throughout callees of _mongocrypt_marking_to_ciphertext.

File: src/mongocrypt-marking.c (hunk 2 above)
Comment (sgolemon) [reply src/mongocrypt-marking.c]: Done

File: src/mongocrypt-marking.c (hunk 5 above)
Comment (sgolemon) [reply src/mongocrypt-marking.c]: As above, done.

**Review by erwee — APPROVED** [review]

**Comment by sgolemon** [comment]
Pushed as 300224f3f1351d2b43de59b96ac681f46de51342

**Pull request closed without merging**